pkg/kv/kvserver/logstore/sideload_disk.go

```diff
@@ -102,15 +99,28 @@ func (ss *DiskSideloadStorage) Put(
 		} else if !oserror.IsNotExist(err) {
 			return err
 		}
-		// createDir() ensures ss.dir exists but will not create any subdirectories
-		// within ss.dir because filename() does not make subdirectories in ss.dir.
-		if err := ss.createDir(); err != nil {
+		// Ensure that ss.dir exists. The filename() is placed directly in ss.dir,
+		// so the next loop iteration should succeed.
+		if err := mkdirAllAndSyncParents(ss.eng, ss.path[0], ss.path[1:], os.ModePerm); err != nil {
 			return err
 		}
 		continue
 	}
 }
 
+// Sync implements SideloadStorage.
+func (ss *DiskSideloadStorage) Sync() error {
```

Comment:
NB: this method implementation does two syscalls: `OpenDir` and `Sync`.

I considered calling `OpenDir` once, around the time when the directories are opened/created, and storing the handle in the struct. Then `Sync` could reusing this handle, and do just a single syscall.

However, the handle holds a file descriptor. There is one sideloaded storage instance per `Range`, so correspondingly we may have many descriptors open at once. Since the descriptors are a limited resource, I decided against this approach.